### PR TITLE
Upgrade LKG to Fsc 4.0 (Released)

### PR DESCRIPTION
**Pull request by otawfik-ms**

### CODE REVIEW

**Comment by KevinRansom** [comment]
Nice

**Comment by enricosada** [comment]
Two stuff:

- can you rebase this branch against `master` branch? The two merge commit are useless and history is more complicated if the merge is done with github button. History hard to follow was  one of the arguments against github merge button.  You can rebase and `push --force` to this branch to overwrite, no need to open another pr, it's fast. Or new branch and cherry pick and push force

- can we use this occasion to use a nuget package for lkg? This continue to add binaries to repo, and size grow. More size, more download time for build ( appveyor etc ). And it's not possibile to shrink after without rewrite history.  A simple nuget created from binaries it's only an x ml file. I can write that, but i think you need to publish it. 
The build already restore from nuget before build.

File: src/FSharpSource.targets [hunk 1]
```diff
@@ -650,6 +650,13 @@
   </PropertyGroup>
 
   <Choose>
+    <When Condition="'$(BuildWith)' == 'LKG'">
+      <PropertyGroup>
+         <FSharpTargetsPath>..\lkg\FSharp-$(LkgVersion)\bin\Microsoft.FSharp.Targets</FSharpTargetsPath>
```
Comment (enricosada) [inline src/FSharpSource.targets]: Why not `$(FSharpSourcesRoot)` here?

File: src/FSharpSource.targets [hunk 2]
```diff
@@ -665,6 +672,18 @@
         <FSharpTargetsPath>..\Proto\$(protoCLIDir)\bin\Microsoft.FSharp-proto.Targets</FSharpTargetsPath>
       </PropertyGroup>
     </When>
+    <When Condition="Exists('$(MSBuildExtensionsPath32)\..\Microsoft SDKs\F#\4.1\Framework\v4.0\Microsoft.FSharp.Targets') AND '$(BuildWith)' == 'LKG'">
+      <PropertyGroup>
+         <FSharpTargetsPath>$(MSBuildExtensionsPath32)\..\Microsoft SDKs\F#\4.1\Framework\v4.0\Microsoft.FSharp.Targets</FSharpTargetsPath>
+         <FSCoreLKGPath Condition="Exists('$(MSBuildExtensionsPath32)\..\Reference Assemblies\Microsoft\FSharp\.NETFramework\v4.0\4.4.1.0\FSharp.Core.dll')">$(MSBuildExtensionsPath32)\..\Reference Assemblies\Microsoft\FSharp\.NETFramework\v4.0\4.4.1.0\FSharp.Core.dll</FSCoreLKGPath>
+      </PropertyGroup>
+    </When>
+    <When Condition="Exists('$(MSBuildExtensionsPath32)\..\Microsoft SDKs\F#\4.0\Framework\v4.0\Microsoft.FSharp.Targets') AND '$(BuildWith)' == 'LKG'">
```
Comment (enricosada) [inline src/FSharpSource.targets]: Is 4.0 exists should be used instead of 4.1? Why

File: src/FSharpSource.targets [hunk 3]
```diff
@@ -665,6 +672,18 @@
...
+      <PropertyGroup>
+         <FSharpTargetsPath>$(MSBuildExtensionsPath32)\..\Microsoft SDKs\F#\4.0\Framework\v4.0\Microsoft.FSharp.Targets</FSharpTargetsPath>
+         <FSCoreLKGPath Condition="Exists('$(MSBuildExtensionsPath32)\..\Reference Assemblies\Microsoft\FSharp\.NETFramework\v4.0\4.4.0.0\FSharp.Core.dll')">$(MSBuildExtensionsPath32)\..\Reference Assemblies\Microsoft\FSharp\.NETFramework\v4.0\4.4.0.0\FSharp.Core.dll</FSCoreLKGPath>
+      </PropertyGroup>
+    </When>
     <When Condition="Exists('$(MSBuildExtensionsPath32)\..\Microsoft SDKs\F#\3.1\Framework\v4.0\Microsoft.FSharp.Targets') AND '$(BuildWith)' == 'LKG'">
       <PropertyGroup>
```
Comment (enricosada) [inline src/FSharpSource.targets]: It's not that the problem we got in appeveyor using 3.1 instead of lkg? I think we should remo these lines, if build == lkg, local lkg inside repo **must** be used

File: src/FSharpSource.targets (hunk 2 above)
Comment (enricosada) [inline src/FSharpSource.targets]: Also read my comment below, i think these are a problem

File: src/FSharpSource.targets (hunk 3 above)
Comment (otawfik-ms) [inline src/FSharpSource.targets]: Sorry, I'm not following. The projects specifying <BuildWith>LKG</BuildWith> are the proto compiler projects:
src\fsharp\Fsc-proto\Fsc-proto.fsproj
src\fsharp\FSharp.Build-proto\FSharp.Build-proto.fsproj
src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj
LKG compiler would be used first if the flag exists. Otherwise, it will pick up the latest installed compiler.

File: src/FSharpSource.targets (hunk 2 above)
Comment (otawfik-ms) [inline src/FSharpSource.targets]: Please see below comment.

File: src/FSharpSource.targets (hunk 1 above)
Comment (otawfik-ms) [inline src/FSharpSource.targets]: Will fix it.

Comment (otawfik-ms) [inline src/FSharpSource.targets]: Part of an older commit. Will fix it.

**Comment by KevinRansom** [comment]
I will Rebase when I pull.  We will move to nuget for the next LKG which will be based on the coreclr.

**Comment by enricosada** [comment]
Yes nuget lkg can wait but the nuget for coreclr need also net40 version so i wrote that because was a good occasion, np anyway

**Comment by KevinRansom** [comment]
I will Rebase when I pull.  We will move to nuget for the next LKG which will be based on the coreclr.
